// bespoken-tools/lib/client/bst-proxy.js
"use strict";
const bespoke_client_1 = require("./bespoke-client");
const lambda_server_1 = require("./lambda-server");
const url_mangler_1 = require("./url-mangler");
const bst_config_1 = require("./bst-config");
const global_1 = require("../core/global");
const function_server_1 = require("./function-server");
(function (ProxyType) {
    ProxyType[ProxyType["GOOGLE_CLOUD_FUNCTION"] = 0] = "GOOGLE_CLOUD_FUNCTION";
    ProxyType[ProxyType["HTTP"] = 1] = "HTTP";
    ProxyType[ProxyType["LAMBDA"] = 2] = "LAMBDA";
})(exports.ProxyType || (exports.ProxyType = {}));
var ProxyType = exports.ProxyType;
const DefaultLambdaPort = 10000;
class BSTProxy {
    constructor(proxyType) {
        this.proxyType = proxyType;
        this.bespokenClient = null;
        this.functionServer = null;
        this.lambdaServer = null;
        this.bespokenHost = "proxy.bespoken.tools";
        this.bespokenPort = 5000;
        this.httpDomain = "localhost";
    }
    static http(targetPort) {
        let tool = new BSTProxy(ProxyType.HTTP);
        tool.httpPort = targetPort;
        return tool;
    }
    static lambda(lambdaFile) {
        let tool = new BSTProxy(ProxyType.LAMBDA);
        tool.functionFile = lambdaFile;
        tool.httpPort = DefaultLambdaPort;
        return tool;
    }
    static cloudFunction(functionFile, functionName) {
        let tool = new BSTProxy(ProxyType.GOOGLE_CLOUD_FUNCTION);
        tool.functionFile = functionFile;
        tool.functionName = functionName;
        tool.httpPort = DefaultLambdaPort;
        return tool;
    }
    static urlgen(url) {
        return url_mangler_1.URLMangler.mangle(url, global_1.Global.config().sourceID(), global_1.Global.config().secretKey());
    }
    bespokenServer(host, port) {
        this.bespokenHost = host;
        this.bespokenPort = port;
        return this;
    }
    targetDomain(host) {
        this.httpDomain = host;
        return this;
    }
    port(port) {
        this.httpPort = port;
        return this;
    }
    start(onStarted) {
        bst_config_1.BSTProcess.run(this.httpPort, this.proxyType, process.pid);
        this.bespokenClient = new bespoke_client_1.BespokeClient(global_1.Global.config().secretKey(), this.bespokenHost, this.bespokenPort, this.httpDomain, this.httpPort);
        let callbackCountDown = 1;
        const callback = function () {
            callbackCountDown--;
            if (callbackCountDown === 0 && onStarted !== undefined) {
                onStarted();
            }
        };
        this.bespokenClient.onConnect = callback;
        this.bespokenClient.connect();
        if (this.proxyType === ProxyType.LAMBDA) {
            callbackCountDown++;
            this.lambdaServer = new lambda_server_1.LambdaServer(this.functionFile, this.httpPort);
            this.lambdaServer.start(callback);
        }
        if (this.proxyType === ProxyType.GOOGLE_CLOUD_FUNCTION) {
            callbackCountDown++;
            this.functionServer = new function_server_1.FunctionServer(this.functionFile, this.functionName, this.httpPort);
            this.functionServer.start(callback);
        }
    }
    stop(onStopped) {
        if (this.bespokenClient !== null) {
            this.bespokenClient.shutdown();
        }
        if (this.lambdaServer !== null) {
            this.lambdaServer.stop(onStopped);
        }
        else if (this.functionServer !== null) {
            this.functionServer.stop(onStopped);
        }
        else {
            if (onStopped !== undefined && onStopped !== null) {
                onStopped();
            }
        }
    }
}
exports.BSTProxy = BSTProxy;
