// bespoken-tools/lib/client/bst-config.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator.throw(value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments)).next());
    });
};
const fs = require("fs");
const logging_helper_1 = require("../core/logging-helper");
const lambda_config_1 = require("./lambda-config");
const source_name_generator_1 = require("../external/source-name-generator");
const spokes_1 = require("../external/spokes");
const Logger = "CONFIG";
const BSTDirectoryName = ".bst";
class BSTConfig {
    constructor() {
        this.configuration = null;
        this.process = null;
    }
    static load() {
        return __awaiter(this, void 0, Promise, function* () {
            yield BSTConfig.bootstrapIfNeeded();
            let data = fs.readFileSync(BSTConfig.configPath());
            let config = JSON.parse(data.toString());
            let bstConfig = new BSTConfig();
            bstConfig.loadFromJSON(config);
            return bstConfig;
        });
    }
    save() {
        BSTConfig.saveConfig(this.configuration);
    }
    nodeID() {
        return this.configuration.nodeID;
    }
    sourceID() {
        return this.configuration.sourceID;
    }
    secretKey() {
        return this.configuration.secretKey;
    }
    applicationID() {
        return this.configuration.applicationID;
    }
    updateApplicationID(applicationID) {
        this.configuration.applicationID = applicationID;
        this.commit();
    }
    commit() {
        let configBuffer = new Buffer(JSON.stringify(this.configuration, null, 4) + "\n");
        fs.writeFileSync(BSTConfig.configPath(), configBuffer);
    }
    loadFromJSON(config) {
        this.configuration = config;
    }
    static configDirectory() {
        return getUserHome() + "/" + BSTDirectoryName;
    }
    static configPath() {
        return BSTConfig.configDirectory() + "/config";
    }
    static bootstrapIfNeeded() {
        return __awaiter(this, void 0, Promise, function* () {
            let directory = BSTConfig.configDirectory();
            if (!fs.existsSync(directory)) {
                fs.mkdirSync(directory);
            }
            if (!fs.existsSync(BSTConfig.configPath())) {
                logging_helper_1.LoggingHelper.info(Logger, "No configuration. Creating one: " + BSTConfig.configPath());
                let configJSON = yield BSTConfig.createConfig();
                BSTConfig.saveConfig(configJSON);
            }
            else {
                let data = fs.readFileSync(BSTConfig.configPath());
                let config = JSON.parse(data.toString());
                if (!config.sourceID) {
                    const pipeConfig = yield BSTConfig.createConfig();
                    config.sourceID = pipeConfig.sourceID;
                    config.secretKey = pipeConfig.secretKey;
                    BSTConfig.saveConfig(config);
                }
            }
        });
    }
    static saveConfig(config) {
        let configBuffer = new Buffer(JSON.stringify(config, null, 4) + "\n");
        fs.writeFileSync(BSTConfig.configPath(), configBuffer);
    }
    static createConfig() {
        return __awaiter(this, void 0, Promise, function* () {
            let lambdaConfig = lambda_config_1.LambdaConfig.defaultConfig().lambdaDeploy;
            const pipeInfo = yield BSTConfig.createSpokesPipe();
            return {
                "sourceID": pipeInfo.endPoint.name,
                "secretKey": pipeInfo.uuid,
                "lambdaDeploy": lambdaConfig
            };
        });
    }
    static createSpokesPipe() {
        return __awaiter(this, void 0, Promise, function* () {
            let isUUIDUnassigned = false;
            let sourceNameGenerator = null;
            let spokesClient = null;
            let spokesPipe = null;
            try {
                sourceNameGenerator = new source_name_generator_1.SourceNameGenerator();
                const generatedKey = yield sourceNameGenerator.callService();
                spokesClient = new spokes_1.SpokesClient(generatedKey.id, generatedKey.secretKey);
                isUUIDUnassigned = yield spokesClient.verifyUUIDisNew();
                if (isUUIDUnassigned) {
                    spokesPipe = yield spokesClient.createPipe();
                }
            }
            catch (error) {
                logging_helper_1.LoggingHelper.error(Logger, "Error : " + error.stack);
                throw Error("Unable to create spokes connection");
            }
            return spokesPipe;
        });
    }
}
exports.BSTConfig = BSTConfig;
class BSTProcess {
    constructor() {
    }
    static running() {
        let process = null;
        if (fs.existsSync(BSTProcess.processPath())) {
            let data = fs.readFileSync(BSTProcess.processPath());
            let json = JSON.parse(data.toString());
            if (BSTProcess.isRunning(json.pid)) {
                process = new BSTProcess();
                process.loadJSON(json);
            }
        }
        return process;
    }
    static isRunning(pid) {
        try {
            process.kill(pid, 0);
            return true;
        }
        catch (e) {
            return e.code === "EPERM";
        }
    }
    static processPath() {
        return getUserHome() + "/" + BSTDirectoryName + "/process";
    }
    static run(port, proxyType, pid) {
        let process = new BSTProcess();
        process.port = port;
        process.proxyType = proxyType;
        process.pid = pid;
        let json = process.json();
        let jsonBuffer = new Buffer(JSON.stringify(json, undefined, 4) + "\n");
        fs.writeFileSync(BSTProcess.processPath(), jsonBuffer);
        return process;
    }
    kill() {
        try {
            process.kill(this.pid, "SIGKILL");
            return true;
        }
        catch (e) {
            console.error("Error killing process[" + this.pid + "] Message: " + e.message);
            return false;
        }
    }
    loadJSON(json) {
        this.port = json.port;
        this.proxyType = json.proxyType;
        this.pid = json.pid;
    }
    json() {
        return {
            "port": this.port,
            "type": this.proxyType,
            "pid": this.pid
        };
    }
}
exports.BSTProcess = BSTProcess;
function getUserHome() {
    return process.env[(process.platform === "win32") ? "USERPROFILE" : "HOME"];
}
