// bespoken-tools/lib/core/webhook-request.js
"use strict";
const querystring = require("querystring");
const buffer_util_1 = require("./buffer-util");
class WebhookRequest {
    constructor(sourceSocket) {
        this.sourceSocket = sourceSocket;
        this.queryParameters = {};
        this.rawContents = new Buffer("");
        this.body = "";
        this.requestID = new Date().getTime();
        if (this.sourceSocket === undefined) {
            this.sourceSocket = null;
        }
    }
    static fromString(sourceSocket, payload, id) {
        let webhookRequest = new WebhookRequest(sourceSocket);
        webhookRequest.append(buffer_util_1.BufferUtil.fromString(payload));
        webhookRequest.requestID = id;
        return webhookRequest;
    }
    append(data) {
        this.rawContents = Buffer.concat([this.rawContents, data]);
        if (this.headers == null) {
            this.headers = {};
            let contentsString = this.rawContents.toString();
            let endIndex = contentsString.indexOf("\r\n\r\n");
            if (endIndex !== -1) {
                this.parseHeaders(contentsString.substr(0, endIndex));
                if (endIndex + 4 < contentsString.length) {
                    let bodyPart = contentsString.substr((endIndex + 4));
                    this.appendBody(bodyPart);
                }
            }
        }
        else {
            this.appendBody(data.toString());
        }
    }
    appendBody(bodyPart) {
        this.body += bodyPart;
    }
    done() {
        if (this.method === "GET") {
            return true;
        }
        return (this.body.length === this.contentLength());
    }
    contentLength() {
        let contentLength = -1;
        if (this.headers != null) {
            let contentLengthString = this.headers["content-length"];
            contentLength = parseInt(contentLengthString);
        }
        return contentLength;
    }
    isPing() {
        return (this.uri.indexOf("/ping") !== -1);
    }
    parseHeaders(headersString) {
        let lines = headersString.split("\n");
        let requestLine = lines[0];
        let requestLineParts = requestLine.split(" ");
        this.method = requestLineParts[0];
        this.uri = requestLineParts[1];
        if (this.uri.indexOf("?") >= 0) {
            const qs = this.uri.replace(/^.*\?/, "");
            this.queryParameters = querystring.parse(qs);
        }
        for (let i = 1; i < lines.length; i++) {
            let headerLine = lines[i];
            let headerParts = headerLine.split(":");
            let key = headerParts[0].toLowerCase();
            this.headers[key] = headerParts[1].trim();
        }
    }
    nodeID() {
        let nodeValue = null;
        if ("node-id" in this.queryParameters) {
            nodeValue = this.queryParameters["node-id"];
        }
        return nodeValue;
    }
    toTCP() {
        return this.rawContents.toString();
    }
    toString() {
        return this.method + " " + this.uri;
    }
    id() {
        return this.requestID;
    }
}
exports.WebhookRequest = WebhookRequest;
